Move TensorOptions ops to c10

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39611 Unify TensorOptions signatures
* **#39492 Move TensorOptions ops to c10**
* #40109 Upgrade msvc to 14.13

This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

smessmer committed:
Move TensorOptions ops to c10



Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

ezyang commented:
This PR is pretty long, can we get a longer PR description?

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -177,6 +178,7 @@
      CUDA: masked_scale_cuda
 
 - func: _sobol_engine_draw(Tensor quasi, int n, Tensor sobolstate, int dimension, int num_generated, ScalarType? dtype) -> (Tensor, Tensor)
+  use_c10_dispatcher: full
```
somebody adding a new op shouldn't have to know about this flag. Adding a new op without specifying it should work. This is only true for all cases if the default is to disable the flag.

ailzhang commented:
@smessmer  I'm still seeing `TensorOptions` showing up for some ops in `RegistrationDeclarations.h`, is this expected? 
```

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

ezyang commented:
> Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take TensorOptions) an creates a wrapper kernel for it that takes the scattered optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool> instead.

What are the conditions under which this wrapper can be removed? Can this be documented in the code? Why can the removal not be done now?

ezyang commented:
I happen to know what `schema_order_arguments` is because we talked about this in personal chat, but I would have expected the PR description / code comments to discuss what is going on with this option.

ezyang commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1518,7 +1518,7 @@ template <typename T>
   static TypePtr call() {
     if (!isCustomClassRegistered<T>()) {
-      throw c10::Error("Type could not be converted to any of the known types.", "");
+      throw c10::Error(std::string("Type ") + std::string(c10::util::get_fully_qualified_type_name<T>()) + " could not be converted to any of the known types.", "");
```
Wonder why this isn't just using `TORCH_CHECK`

bhosmer replied:
+1

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -997,18 +1055,23 @@ def process_native(option):
 
         option['cpp_signature'] = "{} ({})".format(option['return_type'], ", ".join(option['formals_types']))
+        option['schema_order_cpp_signature'] = option['cpp_signature']\
+            .replace('const TensorOptions &',
```
This looks hella fragile

smessmer replied:
how would you do it instead?

ezyang replied:
Well, typically, you would have done this at the time `cpp_signature` was generated, where you have more structure than just a string that you're regexing over (as you are here.) This might be inconvenient given the code, but it's what I would have expected.

bhosmer replied:
+1, how hard would it be to do this at `cpp_signature` creation time? I guess it's not a dealbreaker but this is kind of an antipattern in codegen (important logic done by downstream mutating declarative structures based on string matching)

smessmer replied:
fixed

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -852,6 +891,20 @@ def format_return_type(return_types):
+        elif schema_order_actual == 'memory_format':
+            return 'c10::impl::process_memory_format(options, memory_format)'
+        else:
+            return schema_order_actual
```
The correctness of this translation CRITICALLY relies on the defaulting of these arguments to be `None`, and this defaulting is not universally true:

```
- func: tril_indices(int row, int col, int offset=0, *, ScalarType? dtype=long, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
  dispatch:
    CPU: tril_indices_cpu
    CUDA: tril_indices_cuda
```

See also:

```
- func: _empty_per_channel_affine_quantized(int[] size, *, Tensor scales, Tensor zero_points, int axis, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=contiguous_format) -> Tensor
  category_override: factory
  dispatch:
    CPU: empty_per_channel_affine_quantized_other_backends_stub
    QuantizedCPU: empty_per_channel_affine_quantized_cpu
```

@bhosmer and I attempted to fix the MemoryFormat situation in https://github.com/pytorch/pytorch/pull/34251 but IIRC it was not so easy because the dtype defaulting choices here affect how we do Python argument parsing, and presence of the default perturbs JIT analysis.

ezyang replied:
Actually, it might not be as bad as this implies, because I think we may be handling defaulting in the C++ frontend incorrectly anyway. Current handling for `tril_indices` is:

```
CAFFE2_API Tensor tril_indices(int64_t row, int64_t col, int64_t offset=0, const TensorOptions & options=at::kLong);
```

Which would work OK with the code that you've written here (and will give you a strange result if you try to pass in non-default tensor options, as you'll "lose" the kLong default.)

smessmer replied:
hm yeah this sounds bad. We should probably fix the C++ frontend in a follow up.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -455,13 +486,15 @@ def __getitem__(self, x):
 
 FunctionOption = TypedDict('FunctionOption', {
     'actuals': List[str],
+    'schema_order_actuals': List[str],
```
Expecting docs here and below

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:
```diff
@@ -0,0 +1,110 @@
+}
+
+template<class FuncPtr>
+constexpr auto hacky_wrapper_for_legacy_signatures(FuncPtr) {
```
That's basically what `scatter_tensor_options` in line 55 is. The `hacky_wrapper_for_legacy_signatures` is meant to be more broad, I'm hoping to add non-tensoroptions related functionality like wrapping optional<Tensor> into it too.

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

ailzhang reviewed (approved):
Stamping as we have a fix for the XLA failure. Thanks!

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/common_with_cwrap.py:
```diff
@@ -50,11 +51,16 @@ def set_declaration_defaults(declaration):
     if 'options' not in declaration:
-        declaration['options'] = [{'arguments': declaration['arguments']}]
+        declaration['options'] = [{
+            'arguments': copy.deepcopy(declaration['arguments']),
```
Does this one need to be deep copied?

smessmer replied:
yes. The calling code has some other references to this array and modifies it.

bhosmer commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -99,7 +101,8 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 DEFAULT_FUNCTION_REGISTRATION = CodeTemplate("""\
-m.impl("${unqual_operator_name_with_overload}", TORCH_FN(TypeDefault::${type_wrapper_name}));
+m.impl("${unqual_operator_name_with_overload}",
+       c10::impl::hacky_wrapper_for_legacy_signatures(TORCH_FN(TypeDefault::${type_wrapper_name})));
```
It's a bit of a bummer that our nice clean API has so quickly gotten encrusted with this shim, in particular to the extent these calls will serve as example API usage. Ergonomically it's kind of rough (especially since it might do nothing). Two ideas:

1. teach the codegen to only include the wrapper when necessary. I realize this intrudes on the complete decoupling you were going for here, but I think it would be worth it - the API usage will be better exemplified both for legacy and non-legacy cases. I.e., I think the value of complete decoupling is outweighed by the extent to which it obscures the (evolving) relationship between registered and implemented signatures, and by the readability hit. 

To pitch the change another way: we want registrations for kernels that are already in our desired final state to look like it.

2. hide the wrapper behind TORCH_FN, or a macro that wraps TORCH_FN. I think this is less desirable since it simply hides everything, rather than clarifying when the wrapper is needed, but it would fix the readability issue.

We can teach the codegen when to emit the wrapper just by looking for tensor options in the signature, correct? If so, I'd suggest having a function `needs_legacy_wrapper` where any future triggering conditions would also go (per the comment header in above the metaprogramming for this)

bhosmer commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -131,6 +134,20 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 # add non-virtual declaration to Tensor.cpp
 C10_TENSOR_METHOD_DEFINITION = CodeTemplate("""\
 
+// ${schema_string}
```
if the only difference between this template and `UNBOXEDONLY` is the use of `schema_order_{cpp_signature,method_actuals}`, I think it'd be better to 
- keep a single template and feed in different values for `cpp_signature` and `method_actuals` at the call site
- add a comment above the single template noting the variants
Less redundancy, collapses more easily once the variation is no longer necessary, etc.

smessmer replied:
ok I changed it

bhosmer commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -157,6 +174,20 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 # add method definition in Functions.h
 C10_FUNCTION_DEFINITION = CodeTemplate("""\
 
+// ${schema_string}
```
Same suggestion as with `TENSOR_METHOD_DEFINITION` above

bhosmer commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -24,12 +24,27 @@
+
 FUNCTION_DEFINITION = CodeTemplate("""\
 // ${schema_string}
+Tensor ${function_name}(${method_formals}) {
```
Same suggestion as in `function_wrapper` to use different template variables rather than copy pasting the template

bhosmer commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -24,12 +24,27 @@
   m.impl_UNBOXED("aten::${op_name_with_overload_name}", ${function_name});
 """)
 
+FUNCTION_REGISTRATION = CodeTemplate("""\
```
Same suggestion re making hacky_wrapper conditional. Could use the same central discriminant function to decide; this would actually reinforce the difference as a canonical part of codegen logic and clarify where/why it was necessary.

bhosmer commented on aten/src/ATen/gen_backend_select_register.py:
```diff
@@ -56,28 +72,42 @@ def register_backend_select_methods(declarations, template_path, file_manager):
-                func_reg = FUNCTION_REGISTRATION.substitute(schema_string=option['schema_string'],
-                                                            op_name_with_overload_name=op_name_with_overload_name,
-                                                            function_name=name)
+                if option['use_c10_dispatcher'] == 'full':
```
see, all these call sites would collapse into much neater, less redundant form if a single template was used with different variables

bhosmer commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -158,6 +158,7 @@
     CUDA: _cudnn_rnn_backward
 
 - func: _cudnn_init_dropout_state(float dropout, bool train, int dropout_seed, *, ScalarType dtype, Layout layout, Device device, bool pin_memory=False) -> Tensor
+  #use_c10_dispatcher: full
```
maybe just delete? HA PAYBACK 😛

bhosmer reviewed (commented):
Hey, made it down to `native_functions.yaml`, will resume in a bit - meanwhile a few actionable suggestions in line that I'd pitch pretty enthusiastically. :)

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -339,27 +341,42 @@
 
 # ProfiledType templates
-PROFILE_DISPATCH_UNBOXED = CodeTemplate("""\
+UNBOXED_PROFILE_DISPATCH = CodeTemplate("""\
```
Oops, didn't reach this duplication and the one below during the review yesterday. Sorry to pile on the nits but I really think it helps maintainability to factor this kind of stuff

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -706,16 +738,29 @@ def record_function_input_names():
-    )
+    schema_order_profiled_dispatch_args = ['op', 'c10::DispatchKey::Profiler'] + declaration['schema_order_args']
+
+    if declaration['use_c10_dispatcher'] == 'full':
```
Same re duplication

bhosmer commented on tools/autograd/gen_variable_type.py:
```diff
@@ -737,16 +782,30 @@ def emit_trace_body(declaration):
-        trace_dispatch_args=trace_dispatch_args,
-        assign_return_values=assign_return_values,
-    )
+    if declaration['use_c10_dispatcher'] == 'full':
```
same re duplication

bhosmer reviewed (approved):
LG, a few things inline, mostly copy pasting, but the only major question is whether to conditionalize the hacky wrapper use in this PR or a followup - up to you 😁

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:
```diff
@@ -0,0 +1,112 @@
+    return num_tensoroptions_args > 0;
+}
+
+static_assert(has_tensoroptions_arg<int (int64_t, const TensorOptions&)>(), "");
```
sanity checks and they explain what the type trait is doing for readability.

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

smessmer committed:
Update on "Move TensorOptions ops to c10"


This PR adds `use_c10_dispatcher: full` to ops taking `TensorOptions`. To allow this, since the c10 operator library doesn't know about `TensorOptions`, we need to register the operator kernels as `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead, and also call them this way.

Changes:
- Add `use_c10_dispatcher: full` to those ops
- Write hacky_wrapper_for_legacy_signatures which takes an old-style kernel (i.e. one written to take `TensorOptions`) an creates a wrapper kernel for it that takes the scattered `optional<ScalarType>, optional<Device>, optional<Layout>, optional<bool>` instead.
- Change codegen so that all op registrations are wrapped into `hacky_wrapper_for_legacy_signatures`. This is added to all ops but is a no-op if the op doesn't take TensorOptions. This allows us in the future to just change a kernel signature from `TensorOptions` to the scattered version and have it work without having to touch codegen.
- Change codegen so that the frontend calls those operators with expanded arguments instead of with a `TensorOptions` object. This is required because now the kernels are written in this way.

This PR does not remove TensorOptions special cases from codegen, but instead it separates kernels from the codegen/frontend issues. After this, kernels can be worked on separately without having to touch codegen and codegen can be worked on without having to touch kernels.

Codegen diff: P133121032

Differential Revision: [D21581908](https://our.internmc.facebook.com/intern/diff/D21581908/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/op_registration/hacky_wrapper_for_legacy_signatures.h:
```diff
@@ -0,0 +1,113 @@
+    using parameters_before_tensoroptions =
+        guts::typelist::take_t<gathered_parameter_types, tensoroptions_arg_index>;
+    using parameters_after_tensoroptions =
+        guts::typelist::drop_t<gathered_parameter_types, tensoroptions_arg_index + 1>;
```
Not gonna lie: I'm impressed this all works at all! Do you remember if there were any hidden bombshells, e.g., compiler bugs, you had to work around to get this working? (What are the buried skeletons?)
